soco/services.py

```diff
@@ -283,6 +314,71 @@ def unwrap_arguments(xml_response):
             "{http://schemas.xmlsoap.org/soap/envelope/}Body")[0]
         return dict((i.tag, i.text or "") for i in action_response)
 
+    def compose_args(self, action_name, in_arglist, in_argdict):
+        """Compose the argument list from list of tuples or dictionary form.
+
+        Args:
+            action_name (str): The name of the action to be performed.
+            in_arglist (list): Arguments as a list of ``(name, value)``
+                tuples specifying the name of each argument and its value, eg
+                ``[('InstanceID', 0), ('Speed', 1)]``. The value
+                can be a string or something with a string representation.
+            in_argdict (dict): Arguments as a dict, eg
+                ``{'InstanceID': 0, 'Speed': 1}.
+
+        This merges args and kwargs into a single argument list, with
+        ``DEFAULT_ARGS`` as default values.
+        The ``argdict`` takes precedence over the ``arglist``.
+
+        Returns:
+            list: a list of ``(name, value)`` tuples.
+
+        Raises:
+            `AttributeError`: If this service does not support the action.
+            `ValueError`: If the argument lists do not match the action
+                signature.
+        """
+        in_argdict2 = dict(in_arglist or ())
+
+        action = None
```

Comment:
Variables from inside loops retain their value after the loop ends and the for loop has an else clause that gets executed if there was no `break`. Therefore:
```python
action = None
for act in self.actions:
    if act.name == action_name:
        action = act
        break
if action is None:
    raise AttributeError('Unknown Action: {0}'.format(action_name))
```
can be written as:
```python
# Find appropriate action
for action in self.actions:
    if action.name == action_name:
        break
else:
    raise AttributeError('Unknown Action: {0}'.format(action_name))
```
which gets you rid of that nastily abbreviated `act` placeholder, although the retain-value-after-loop-end is lesser known, so better attach a comment

Reply:
Thanks!